crates/bevy_text/src/pipeline.rs

```diff
@@ -1,41 +1,225 @@
-use ab_glyph::{PxScale, ScaleFont};
+use std::sync::{Arc, Mutex};
+
 use bevy_asset::{Assets, Handle, HandleId};
 use bevy_ecs::component::Component;
 use bevy_ecs::system::Resource;
 use bevy_math::Vec2;
 use bevy_render::texture::Image;
 use bevy_sprite::TextureAtlas;
-use bevy_utils::HashMap;
+use bevy_utils::{
+    tracing::{error, info, warn},
+    HashMap,
+};
 
-use glyph_brush_layout::{FontId, GlyphPositioner, SectionGeometry, SectionText};
+use cosmic_text::{Attrs, AttrsList, Buffer, BufferLine, Family, Metrics, Wrap};
 
 use crate::{
-    error::TextError, glyph_brush::GlyphBrush, scale_value, BreakLineOn, Font, FontAtlasSet,
-    FontAtlasWarning, PositionedGlyph, TextAlignment, TextSection, TextSettings, YAxisOrientation,
+    error::TextError, BreakLineOn, Font, FontAtlasSet, FontAtlasWarning, FontRef, PositionedGlyph,
+    TextAlignment, TextSection, TextSettings, YAxisOrientation,
 };
 
-#[derive(Default, Resource)]
-pub struct TextPipeline {
-    brush: GlyphBrush,
-    map_font_id: HashMap<HandleId, FontId>,
+// TODO: cache buffers / store buffers on the entity
+// TODO: reconstruct byte indices
+// TODO: rescale font sizes in all examples
+// TODO: fix any broken examples
+// TODO: solve spans with different font sizes, see https://github.com/pop-os/cosmic-text/issues/64
+// TODO: (future work) split text entities into section entities
+// TODO: (future work) text editing
+// TODO: font validation
+
+// TODO: the only reason we need a mutex is due to TextMeasure
```

Comment:
@TimJentzsch @nicoburns I remember solving similar headaches in taffy. Do either of you have advice here?

Reply:
@nicoburns referenced [in discord](https://discord.com/channels/691052431525675048/743663673393938453/1114175469466693644) this PR https://github.com/DioxusLabs/taffy/pull/490 which could in theory remove the need for mutexes.

Reply:
Yes, duplicating my comment from Discord to save people clicking through: It potentially helps in two ways:
- It makes the `Send+Sync` bound on measure functions optional (
- It allows measure functions to (mutably if need be) borrow from a global context which is passed in to the compute_layout function

I think it is mostly the latter that will help here, as allows the FontSystem to be stored outside of Taffy and borrowed by measure functions (via their new context parameter).

Reply:
#490 has landed in Taffy main (using a slightly different approach: a value of a user-defined context type is now stored on each node, and a single measure function is passed into `compute_layout` (which can borrow external variables as needed, and only lives as long as a single layout run avoiding most lifetime issues).